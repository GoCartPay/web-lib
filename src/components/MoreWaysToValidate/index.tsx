import React, { useState } from 'react';
import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';
import CircularProgress from '@mui/material/CircularProgress';
import Alert from '@mui/material/Alert';
import CheckBoxIcon from '@mui/icons-material/CheckBox';
import { keyframes } from '@emotion/react';

import { BigButton } from '../BigButton';
import RadioGroup from '../RadioGroup';
import { Radio } from '../RadioGroup';

const fadeInUp = keyframes`
  from {
    opacity: 0;
    transform: translate3d(0, 100%, 0);
  }
  to {
    opacity: 1;
    transform: translate3d(0, 0, 0);
  }
`

const fadeOutDown = keyframes`
  from {
    opacity: 1;
  }
  to {
    opacity: 0;
    transform: translate3d(0, 100%, 0);
  }
`


type MoreWaysToValidateProps = {
  /** whether loading icon should be displayed inside main button */
  isLoading?: boolean
  /** will show or hide component. component will fade in and out */
  isOpen: boolean
  /** callback func that fires when user clicks on enter new code*/
  onEnterNewCode: () => void
  /** callback func that fires when user cancels resending auth token */
  onCancel: () => void
  /** callback func that fires when user resends auth, gives access to auth type they choose */
  onResend: (otpMethod: string) => void
  /** radio options that get passed into radio group. These should have both a label and a string */
  radioOptions: Radio[]
  /** toggles whether to show success message */
  showSuccess?: boolean
  /** the success message to display to the user when showSuccess is true */
  successMessage: string
};

/**
 * Try Again OTP prompt to use when user wants to select a new way to verify or resend OTP code
 * @param {MoreWaysToValidateProps} MoreWaysToValidateProps Component properties
 * @returns {JSX.Element}
 */
const MoreWaysToValidate: React.FC<MoreWaysToValidateProps> = ({ 
  isLoading = false,
  isOpen,
  onEnterNewCode,
  onCancel,
  onResend,
  radioOptions, 
  showSuccess = false,
  successMessage,
}) => {

  const [selectedOTPMethod, setSelectedOTPMethod] = useState(radioOptions[0].value);
  const [fadeIn, setFadeIn] = useState(true);

  const handleRadioBtnChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSelectedOTPMethod((event.target as HTMLInputElement).value)
  };

  const handleOnExit = () => setFadeIn(false);

  const handleOnCancel= () => { 
    handleOnExit();
    setTimeout(() => {onCancel()}, 400);
  };

  const handleResend = () => onResend(selectedOTPMethod);

  const handleEnterNewCode = () => {
    setTimeout(() => { onEnterNewCode()}, 400);
  };

  return isOpen ? (
    <Box
      sx={{ animation: `${fadeIn ? `${fadeInUp} 400ms` : `${fadeOutDown} 400ms`}` }}
    >
      <Typography variant='body1' sx={{ opacity: 0.6, mb: 2 }}>
        Where would you like to receive a new validation code?
      </Typography>
      <RadioGroup
        value={selectedOTPMethod}
        onChange={handleRadioBtnChange}
        radioOptions={radioOptions}
      />
      <Box sx={{ mt: 4, borderTop: 1, borderColor: '#DCDEE5', pt: 3.5 }}>
        {showSuccess &&
          <Box sx={{ my: 1.5 }}>
            <Alert
              severity="success"
              icon={<CheckBoxIcon sx={{ height: 32, width: 32 }} />}
              sx={{ borderRadius: '8px', height: '56px', display: 'flex', alignItems: 'center', lineHeight: '56px', fontWeight: 600 }}
            >
              {successMessage}
            </Alert>
          </Box>
        }
        <Box sx={{ my: 1.5 }}>
          <BigButton 
            variant="contained" 
            sx={{ height: '56px !important' }} 
            labelText={showSuccess ? 'Enter new code' : 'Resend code'} 
            onClick={showSuccess ?  handleEnterNewCode : handleResend}
          >
            {isLoading && <CircularProgress sx={{ color: '#ffffff' }} />}
          </BigButton>
        </Box>
        <Box sx={{ my: 1.5 }}>
          { !showSuccess &&
             <BigButton 
              variant="outlined" 
              sx={{ height: '56px !important' }} 
              labelText='Cancel' 
              onClick={handleOnCancel} 
            />
          }
        </Box>
      </Box>
    </Box>
  ) : null;
};

export default MoreWaysToValidate;