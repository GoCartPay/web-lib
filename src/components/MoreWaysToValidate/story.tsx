import React, { useState } from 'react';
import MoreWaysToValidate from '.';

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: 'Spruce/Components/MoreWaysToValidate',
  component: MoreWaysToValidate,
}

const userPhone = '(•••) ••••• 554';
const userEmail = 'p••l@philj.com';

const radioOptions = [
    { value: 'SMS', label: `Send to ${userPhone}` },
    { value: 'email', label: `Send to ${userEmail}`}
];

const Template = () => {
  const [isOpen, setIsOpen] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [resendSuccess, setResendSuccess] = useState(false);
  const [authType, setAuthType] = useState('SMS');

  const handleResendOTP = (val: string) => {
    setIsLoading(true);
    setAuthType(val);
    setTimeout(() => {
      setIsLoading(false);
      setResendSuccess(true);
    },1000);
  };

  const handleEnterNewCode = () => {
    setIsOpen(false);
    setResendSuccess(false);
  };

  if (!isOpen) return <button onClick={() => setIsOpen(true)}>Open</button>;

  return (
    <div style={{ maxWidth: '327px', margin: '0px auto' }}>
      <MoreWaysToValidate
        isLoading={isLoading}
        isOpen={isOpen}
        radioOptions={radioOptions}
        onCancel={() => setIsOpen(false)}
        onEnterNewCode={handleEnterNewCode}
        onResend={handleResendOTP}
        showSuccess={resendSuccess}
        successMessage={`Sent to ${authType === 'SMS' ? userPhone : userEmail}`}
      />
    </div>
  )
};

export const Default = Template.bind({})