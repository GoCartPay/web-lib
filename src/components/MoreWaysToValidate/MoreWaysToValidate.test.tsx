import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

import MoreWaysToValidate from './index';

const userPhone = 'xxxxxxxxxx';
const userEmail = 'gocart@gocart.com';
const successMessage = 'this is a success message';

const radioOptions = [
  { value: 'SMS', label: `Send to ${userPhone}` },
  { value: 'email', label: `Send to ${userEmail}` }
];

const mockEnterCode = jest.fn();
const mockCancel = jest.fn();
const mockResend = jest.fn();

const Base = (props: any) => { 
  
  const baseProps = {
    radioOptions,
    isOpen: true,
    onEnterNewCode: mockEnterCode,
    onCancel: mockCancel, 
    onResend: mockResend,
    successMessage: successMessage
  };
  const mergedProps = {...baseProps, ...props};

  return (
  <MoreWaysToValidate
    {...mergedProps}
  />
);};


describe('More Ways to Validate Screen behaves correctly', () => {

  it('renders without crashing and matches snapshot', () => {
    const { container } = render(
      <Base />
    )

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not render when isOpen is false', () => {
    const { container } = render(
      <Base isOpen={false} />
    );

    expect(container.firstChild).toBeNull();
  });

  it('onResend, onCancel callbacks should fire correctly', () => {
    render(
      <Base/>
    );

    const resendBtn = screen.getByRole('button', { name: 'Resend code'});
    const cancelBtn = screen.getByRole('button', { name: 'Cancel'});

    userEvent.click(resendBtn);
    expect(mockResend).toBeCalledTimes(1);

    userEvent.click(cancelBtn);
    expect(mockCancel).toHaveBeenCalledTimes(0);
    setTimeout( () => { expect(mockCancel).toBeCalledTimes(1) }, 400);

  });

  it('onEnterNewCode callback should fire correctly', () => {

    render(
      <Base 
        showSuccess
      />
    );
    
    const reenterBtn = screen.getByRole('button', { name: 'Enter new code'});

    userEvent.click(reenterBtn);
    expect(mockResend).toBeCalledTimes(1);
  })

  it('should should a success message when success happens', () => {
    render(
      <Base 
        showSuccess
        successMessage={successMessage}
      />
    );

    const alert = screen.getByRole('alert');
    expect(alert).toHaveTextContent(successMessage)
  })

  it('shows the correct layout when success is true', () => {
   const { container } = render(
    <Base 
      showSuccess
      successMessage={successMessage}
    />
   );

   expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a radio button for each radio option passed in', () => {
    render(
      <Base />
    );
    
    expect(screen.getAllByRole('radio').length).toEqual(radioOptions.length);
  });
})