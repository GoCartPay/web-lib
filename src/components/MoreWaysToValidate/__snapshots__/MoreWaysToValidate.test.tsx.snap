// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`More Ways to Validate Screen behaves correctly renders without crashing and matches snapshot 1`] = `
<div
  class="MuiBox-root css-14ft4zl"
>
  <p
    class="MuiTypography-root MuiTypography-body1 css-lwiwm9-MuiTypography-root"
  >
    Where would you like to receive a new validation code?
  </p>
  <div
    class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
    role="radiogroup"
  >
    <div
      class="MuiBox-root css-pb69ky"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth css-b5eqg5-MuiFormControl-root"
      >
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
        >
          <span
            class="MuiRadio-root MuiRadio-colorPrimary MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root Mui-checked css-vqmohf-MuiButtonBase-root-MuiRadio-root"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input css-1m9pwf3"
              name="mui-1"
              type="radio"
              value="SMS"
            />
            <span
              class="css-hyxlzm"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                data-testid="RadioButtonUncheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-11zohuh-MuiSvgIcon-root"
                data-testid="RadioButtonCheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <p
            class="MuiTypography-root MuiTypography-body2 css-hoiwr4-MuiTypography-root"
          >
            Send to xxxxxxxxxx
          </p>
        </label>
      </div>
    </div>
    <div
      class="MuiBox-root css-pb69ky"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth css-l78efq-MuiFormControl-root"
      >
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
        >
          <span
            class="MuiRadio-root MuiRadio-colorPrimary MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root css-vqmohf-MuiButtonBase-root-MuiRadio-root"
          >
            <input
              class="PrivateSwitchBase-input css-1m9pwf3"
              name="mui-1"
              type="radio"
              value="email"
            />
            <span
              class="css-hyxlzm"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                data-testid="RadioButtonUncheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hhw7if-MuiSvgIcon-root"
                data-testid="RadioButtonCheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <p
            class="MuiTypography-root MuiTypography-body2 css-c9fm5x-MuiTypography-root"
          >
            Send to gocart@gocart.com
          </p>
        </label>
      </div>
    </div>
  </div>
  <div
    class="MuiBox-root css-g0d7gu"
  >
    <div
      class="MuiBox-root css-10khgmf"
    >
      <div
        class="MuiBox-root css-w5s6vd"
      >
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-yt3qrx-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Resend code
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiBox-root css-10khgmf"
    >
      <div
        class="MuiBox-root css-6p8c8m"
      >
        <button
          class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-154jra7-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Cancel
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`More Ways to Validate Screen behaves correctly shows the correct layout when success is true 1`] = `
<div
  class="MuiBox-root css-14ft4zl"
>
  <p
    class="MuiTypography-root MuiTypography-body1 css-lwiwm9-MuiTypography-root"
  >
    Where would you like to receive a new validation code?
  </p>
  <div
    class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
    role="radiogroup"
  >
    <div
      class="MuiBox-root css-pb69ky"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth css-b5eqg5-MuiFormControl-root"
      >
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
        >
          <span
            class="MuiRadio-root MuiRadio-colorPrimary MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root Mui-checked css-vqmohf-MuiButtonBase-root-MuiRadio-root"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input css-1m9pwf3"
              name="mui-5"
              type="radio"
              value="SMS"
            />
            <span
              class="css-hyxlzm"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                data-testid="RadioButtonUncheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-11zohuh-MuiSvgIcon-root"
                data-testid="RadioButtonCheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <p
            class="MuiTypography-root MuiTypography-body2 css-hoiwr4-MuiTypography-root"
          >
            Send to xxxxxxxxxx
          </p>
        </label>
      </div>
    </div>
    <div
      class="MuiBox-root css-pb69ky"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth css-l78efq-MuiFormControl-root"
      >
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
        >
          <span
            class="MuiRadio-root MuiRadio-colorPrimary MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root css-vqmohf-MuiButtonBase-root-MuiRadio-root"
          >
            <input
              class="PrivateSwitchBase-input css-1m9pwf3"
              name="mui-5"
              type="radio"
              value="email"
            />
            <span
              class="css-hyxlzm"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                data-testid="RadioButtonUncheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hhw7if-MuiSvgIcon-root"
                data-testid="RadioButtonCheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <p
            class="MuiTypography-root MuiTypography-body2 css-c9fm5x-MuiTypography-root"
          >
            Send to gocart@gocart.com
          </p>
        </label>
      </div>
    </div>
  </div>
  <div
    class="MuiBox-root css-g0d7gu"
  >
    <div
      class="MuiBox-root css-10khgmf"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardSuccess MuiAlert-standard css-eldf5y-MuiPaper-root-MuiAlert-root"
        role="alert"
      >
        <div
          class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-56kpne-MuiSvgIcon-root"
            data-testid="CheckBoxIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
        </div>
        <div
          class="MuiAlert-message css-acap47-MuiAlert-message"
        >
          this is a success message
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-10khgmf"
    >
      <div
        class="MuiBox-root css-w5s6vd"
      >
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-yt3qrx-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Enter new code
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiBox-root css-10khgmf"
    />
  </div>
</div>
`;
